src/findone-parallel-performance.c

```diff
@@ -0,0 +1,239 @@
+      MONGOC_ERROR ("database_drop: %s\n", error.message);
+      abort ();
+   }
+   mongoc_database_destroy (db);
+   mongoc_client_pool_push (pool, client);
+   mongoc_uri_destroy (uri);
+}
+
+static void
+findone_parallel_perf_teardown (perf_test_t *test)
+{
+   findone_parallel_perf_test_t *findone_parallel_test =
+      (findone_parallel_perf_test_t *) test;
+
+   mongoc_client_pool_destroy (findone_parallel_test->pool);
+   bson_free (findone_parallel_test->contexts);
+}
+
+static void
+findone_parallel_perf_before (perf_test_t *test)
+{
+   findone_parallel_perf_test_t *findone_parallel_test =
+      (findone_parallel_perf_test_t *) test;
+   int i;
+
+   if (FINDONE_COUNT % findone_parallel_test->n_threads != 0) {
+      MONGOC_ERROR (
+         "FINDONE_COUNT (%d) is not divisible by number of threads: %d",
+         FINDONE_COUNT,
+         findone_parallel_test->n_threads);
+      MONGOC_ERROR ("Consider revising test or using integer division.");
+      abort ();
+   }
+
+   for (i = 0; i < findone_parallel_test->n_threads; i++) {
+      findone_parallel_test->contexts[i].client =
+         mongoc_client_pool_pop (findone_parallel_test->pool);
+      findone_parallel_test->contexts[i].n_findone_to_run =
+         FINDONE_COUNT / findone_parallel_test->n_threads;
+   }
+}
+
+static void
+findone_parallel_perf_after (perf_test_t *test)
+{
+   findone_parallel_perf_test_t *findone_parallel_test =
+      (findone_parallel_perf_test_t *) test;
+   int i;
+
+   if (findone_parallel_test->n_threads > 100) {
+      MONGOC_ERROR ("Error: trying to start test with %d threads.",
+                    findone_parallel_test->n_threads);
+      MONGOC_ERROR ("Cannot start test with n_threads > 100.");
+      MONGOC_ERROR ("libmongoc uses a default maxPoolSize of 100. Cannot pop "
+                    "more than 100.");
+      MONGOC_ERROR ("Consider revising this test to use a larger pool size.");
+      abort ();
+   }
+
+   for (i = 0; i < findone_parallel_test->n_threads; i++) {
+      mongoc_client_pool_push (findone_parallel_test->pool,
+                               findone_parallel_test->contexts[i].client);
+   }
+}
+
+static perf_test_t *
+findone_parallel_perf_new (const char *name, int n_threads)
+{
+   findone_parallel_perf_test_t *findone_parallel_test =
+      bson_malloc0 (sizeof (findone_parallel_perf_test_t));
+   perf_test_t *test = (perf_test_t *) findone_parallel_test;
+   int64_t data_size;
+
+   findone_parallel_test->n_threads = n_threads;
+   data_size = FINDONE_FILTER_SIZE * FINDONE_COUNT;
+
+   perf_test_init (test, name, NULL /* data path */, data_size);
+   test->task = findone_parallel_perf_task;
+   test->setup = findone_parallel_perf_setup;
+   test->teardown = findone_parallel_perf_teardown;
+   test->before = findone_parallel_perf_before;
+   test->after = findone_parallel_perf_after;
+   return test;
+}
+
+void
+findone_parallel_perf (void)
+{
+   perf_test_t *perf_tests[] = {
```

Comment:
This is nice, since it's easy to just "hang" a new test here and have it seamlessly be added. :-)